### PR TITLE
Outflow

### PR DESCRIPTION
**Pull request by tverbeiren**

### CODE REVIEW

**Comment by tverbeiren** [comment]
Command:

```bash
nextflow run target/nextflow/demultiplex/main.nf \
  -resume \
  -with-docker \
  --input "testData/200624_A00834_0183_BHMTFYDRXX.tar.gz" \
  --output "fastq" \
  --output_multiqc "qc/multiqc_report.html" \
  --output_falco "qc/fastqc" \
  --publish_dir "publish_dir"
```

yields:

```
publish_dir
└── 20241010_090429
    ├── fastq
    │   ├── Sample1_S1_L001_R1_001.fastq.gz
    │   ├── Sample23_S3_L001_R1_001.fastq.gz
    │   ├── SampleA_S2_L001_R1_001.fastq.gz
    │   ├── Undetermined_S0_L001_R1_001.fastq.gz
    │   └── sampletest_S4_L001_R1_001.fastq.gz
    ├── logs
    │   ├── Errors.log
    │   ├── FastqComplete.txt
    │   ├── Info.log
    │   └── Warnings.log
    ├── qc
    │   ├── fastqc
    │   │   ├── Sample1_S1_L001_R1_001.fastq.gz_fastqc_data.txt
    │   │   ├── Sample1_S1_L001_R1_001.fastq.gz_fastqc_report.html
    │   │   ├── Sample1_S1_L001_R1_001.fastq.gz_summary.txt
    │   │   ├── Undetermined_S0_L001_R1_001.fastq.gz_fastqc_data.txt
    │   │   ├── Undetermined_S0_L001_R1_001.fastq.gz_fastqc_report.html
    │   │   └── Undetermined_S0_L001_R1_001.fastq.gz_summary.txt
    │   └── multiqc_report.html
    └── reports
        ├── Adapter_Cycle_Metrics.csv
        ├── Adapter_Metrics.csv
        ├── Demultiplex_Stats.csv
        ├── Demultiplex_Tile_Stats.csv
        ├── IndexMetricsOut.bin
        ├── Index_Hopping_Counts.csv
        ├── Quality_Metrics.csv
        ├── Quality_Tile_Metrics.csv
        ├── RunInfo.xml
        ├── SampleSheet.csv
        ├── Top_Unknown_Barcodes.csv
        └── fastq_list.csv

7 directories, 28 files
```

**Review by DriesSchaumont — APPROVED** [review]
LGTM!

File: src/demultiplex/main.nf [hunk 1]
```diff
@@ -126,6 +134,11 @@ workflow run_wf {
           "output_multiqc": "output_multiqc"
         ]
       )
+      | map{ id, state -> 
+        def newId = "${date}/".toString()
```
Comment (DriesSchaumont) [inline src/demultiplex/main.nf:138]: Does this not fail when we have multiple runs that are processed at the same time?

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]

File: src/runner/main.nf [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          ]
+        ]
+      )
+
+  emit:
+    output_ch
+}
+
+def get_version(txt) {
+  matcher = txt =~ /(?m)^version:\s+(.*)$/
```
Comment (DriesSchaumont) [inline src/runner/main.nf:62]: I would use a yaml slurper; something like:
```
 def yamlSlurper = new org.yaml.snakeyaml.Yaml()
 def loaded_viash_config = yamlSlurper.load(inputFile)
 loaded_viash_config.version
```
I have not tested this specific code, but in nextflow `org.yaml.snakeyaml.Yaml()` is available

File: src/io/publish/code.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+echo
+echo "par_output:         $par_output"
+echo "par_output_falco:   $par_output_falco"
+echo "par_output_multiqc: $par_output_multiqc"
+
+cp -L "$par_input"/* "$par_output"
+cp -rL "$par_input_falco"/* "$par_output_falco"
+cp -L "$par_input_multiqc" "$par_output_multiqc"
+
```
Comment (DriesSchaumont) [inline src/io/publish/code.sh:15]: Can we make this abstract by letting the publish component take a list of inputs and a list of outputs, and a loop over them?

File: src/runner/nextflow.config [hunk 4]
```diff
@@ -0,0 +1,6 @@
+manifest {
+  nextflowVersion = '!>=20.12.1-edge'
```
Comment (DriesSchaumont) [inline src/runner/nextflow.config:2]: I think we might need a more recent verison.. but I have no clue which one...

File: src/io/publish/config.vsh.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+resources:
+  - type: bash_script
+    path: ./code.sh
+
+engines:
+  - type: docker
+    image: debian:stable-20220912
+    setup:
+      - type: docker
+        run: apt update && apt install -y procps
```
Comment (DriesSchaumont) [inline src/io/publish/config.vsh.yaml:43]:
```suggestion
      - type: apt
        packages:
          - procps        
```

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]

File: src/io/publish/code.sh (hunk 3 above)
Comment (tverbeiren) [reply src/io/publish/code.sh]: This does not work because the output should contain a wildcard character.

File: src/runner/nextflow.config (hunk 4 above)
Comment (tverbeiren) [reply src/runner/nextflow.config:2]: I suggest to leave this for now.

Comment (DriesSchaumont) [reply src/runner/nextflow.config:2]: Sure.

**Review by DriesSchaumont — APPROVED** [review]
LGTM!

**Pull request merged**